drivers/disk/nvme/nvme.h

```diff
@@ -0,0 +1,446 @@
+
+	/** Atomic Compare & Write Unit */
+	uint16_t		acwu;
+	uint16_t		reserved6;
+
+	/** SGL Support */
+	uint32_t		sgls;
+
+	/** Maximum Number of Allowed Namespaces */
+	uint32_t		mnan;
+
+	/* bytes 540-767: Reserved */
+	uint8_t			reserved7[224];
+
+	/** NVM Subsystem NVMe Qualified Name */
+	uint8_t			subnqn[256];
+
+	/* bytes 1024-1791: Reserved */
+	uint8_t			reserved8[768];
+
+	/* bytes 1792-2047: NVMe over Fabrics specification */
+	uint8_t			reserved9[256];
+
+	/* bytes 2048-3071: power state descriptors */
+	struct nvme_power_state power_state[32];
+
+	/* bytes 3072-4095: vendor specific */
+	uint8_t			vs[1024];
+} __packed __aligned(4);
+
+static inline
+void nvme_controller_data_swapbytes(struct nvme_controller_data *s)
+{
+#if _BYTE_ORDER != _LITTLE_ENDIAN
+	s->vid = sys_le16_to_cpu(s->vid);
+	s->ssvid = sys_le16_to_cpu(s->ssvid);
+	s->ctrlr_id = sys_le16_to_cpu(s->ctrlr_id);
+	s->ver = sys_le32_to_cpu(s->ver);
+	s->rtd3r = sys_le32_to_cpu(s->rtd3r);
+	s->rtd3e = sys_le32_to_cpu(s->rtd3e);
+	s->oaes = sys_le32_to_cpu(s->oaes);
+	s->ctratt = sys_le32_to_cpu(s->ctratt);
+	s->rrls = sys_le16_to_cpu(s->rrls);
+	s->crdt1 = sys_le16_to_cpu(s->crdt1);
+	s->crdt2 = sys_le16_to_cpu(s->crdt2);
+	s->crdt3 = sys_le16_to_cpu(s->crdt3);
+	s->oacs = sys_le16_to_cpu(s->oacs);
+	s->wctemp = sys_le16_to_cpu(s->wctemp);
+	s->cctemp = sys_le16_to_cpu(s->cctemp);
+	s->mtfa = sys_le16_to_cpu(s->mtfa);
+	s->hmpre = sys_le32_to_cpu(s->hmpre);
+	s->hmmin = sys_le32_to_cpu(s->hmmin);
+	s->rpmbs = sys_le32_to_cpu(s->rpmbs);
+	s->edstt = sys_le16_to_cpu(s->edstt);
+	s->kas = sys_le16_to_cpu(s->kas);
+	s->hctma = sys_le16_to_cpu(s->hctma);
+	s->mntmt = sys_le16_to_cpu(s->mntmt);
+	s->mxtmt = sys_le16_to_cpu(s->mxtmt);
+	s->sanicap = sys_le32_to_cpu(s->sanicap);
+	s->hmminds = sys_le32_to_cpu(s->hmminds);
+	s->hmmaxd = sys_le16_to_cpu(s->hmmaxd);
+	s->nsetidmax = sys_le16_to_cpu(s->nsetidmax);
+	s->endgidmax = sys_le16_to_cpu(s->endgidmax);
+	s->anagrpmax = sys_le32_to_cpu(s->anagrpmax);
+	s->nanagrpid = sys_le32_to_cpu(s->nanagrpid);
+	s->pels = sys_le32_to_cpu(s->pels);
+	s->maxcmd = sys_le16_to_cpu(s->maxcmd);
+	s->nn = sys_le32_to_cpu(s->nn);
+	s->oncs = sys_le16_to_cpu(s->oncs);
+	s->fuses = sys_le16_to_cpu(s->fuses);
+	s->awun = sys_le16_to_cpu(s->awun);
+	s->awupf = sys_le16_to_cpu(s->awupf);
+	s->acwu = sys_le16_to_cpu(s->acwu);
+	s->sgls = sys_le32_to_cpu(s->sgls);
+	s->mnan = sys_le32_to_cpu(s->mnan);
+#else
+	ARG_UNUSED(s);
+#endif
+}
+
+#include <zephyr/device.h>
+#include <zephyr/drivers/pcie/pcie.h>
+#include <zephyr/drivers/pcie/msi.h>
+
+#define NVME_PCIE_BAR_IDX 0
+
+#define NVME_REQUEST_AMOUNT (CONFIG_NVME_ADMIN_ENTRIES +	\
+			     CONFIG_NVME_IO_ENTRIES)
+
+/* admin queue + io queue(s) */
+#define NVME_PCIE_MSIX_VECTORS 1 + CONFIG_NVME_IO_QUEUES
+
+#define NVME_QUEUE_ALLOCATE(name, n_entries)				\
+	static struct nvme_command cmd_##name[n_entries] __aligned(0x1000); \
+	static struct nvme_completion cpl_##name[n_entries] __aligned(0x1000); \
```

Comment:
Is this alignment required for _all_ platforms or is it a specific requirement of the specific platform that implemented this driver?